File: server/app/models/service_hours.py
"""
Service hours ORM model.
"""
from typing import Any

from sqlalchemy import Index, String, UniqueConstraint
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Mapped, mapped_column

from app.db.session import Base
from app.models.base import TimestampMixin


class ServiceHours(Base, TimestampMixin):
    __tablename__ = "service_hours"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    tenant_id: Mapped[str] = mapped_column(String(32), nullable=False)
    name: Mapped[str] = mapped_column(String(64), nullable=False)
    description: Mapped[str | None] = mapped_column(String(256), nullable=True)
    timezone: Mapped[str] = mapped_column(
        String(64), nullable=False, server_default="Asia/Shanghai"
    )
    weekly_periods: Mapped[list[dict[str, Any]]] = mapped_column(
        JSONB, nullable=False, server_default="[]"
    )
    holidays: Mapped[list[dict[str, Any]]] = mapped_column(
        JSONB, nullable=False, server_default="[]"
    )
    makeup_days: Mapped[list[dict[str, Any]]] = mapped_column(
        JSONB, nullable=False, server_default="[]"
    )

    __table_args__ = (
        Index("ix_service_hours_tenant_id", "tenant_id"),
        UniqueConstraint("tenant_id", "name", name="uq_service_hours_tenant_name"),
    )

File: web/models/service-hours.ts
export type WeeklyServicePeriod = {
  day_of_week: number
  start: string
  end: string
}

export type ServiceHoursDateTimeRange = {
  name: string | null
  start_at: string
  end_at: string
}

export type ServiceHours = {
  id: number
  tenant_id: string
  name: string
  description: string | null
  timezone: string
  weekly_periods: WeeklyServicePeriod[]
  holidays: ServiceHoursDateTimeRange[]
  makeup_days: ServiceHoursDateTimeRange[]
  created_at: string
  updated_at: string
}

export type CreateServiceHoursPayload = {
  name: string
  description?: string | null
  timezone: string
  weekly_periods: WeeklyServicePeriod[]
  holidays: ServiceHoursDateTimeRange[]
  makeup_days: ServiceHoursDateTimeRange[]
}

export type UpdateServiceHoursPayload = Partial<CreateServiceHoursPayload>
